lib/linked_list.rb

```diff
@@ -3,89 +3,174 @@
 class Node
   attr_reader :data # allow external entities to read value but not write
   attr_accessor :next # allow external entities to read or write next node
+  attr_accessor :previous
 
-  def initialize(value, next_node = nil)
+  def initialize(value, next_node = nil, previous_node = nil)
     @data = value
     @next = next_node
+    @previous = previous_node
+
   end
 end
 
 # Defines the singly linked list
 class LinkedList
     def initialize
       @head = nil # keep the head private. Not accessible outside this class
+      @tail = nil 
     end
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      if @head.nil? 
+        @head = @tail = Node.new(value)
+      else 
+        new_node = Node.new(value)
+        new_node.next = @head
+        @head.previous = new_node
+        @head = new_node

```

Comment:
Could be simplified a bit

```suggestion
        @head = Node.new(value, @head)
        @head.next.previous = @head
```